Settings page: Show error message when duplicate tracking selected

acicovic opened the pull request:
## Description
In the settings page, if the user selects the same element in the "Post Types to Track" and "Page Types to Track" sections, an error message will be shown to them telling them that the element should be selected only in one of the sections.

## Motivation and Context
The reasoning behind this is that we should not allow duplicate tracking. This PR partly addresses the second point found in https://github.com/Parsely/wp-parsely/pull/482#issuecomment-963287549. Also related: #483.

## How Has This Been Tested?
- Tested manually using the Parse.ly plugin's options page,
- Added an integration test.

## Screenshots
Error that appears when items are selected in both "Post Types to Track" and "Page Types to Track":

![image](https://user-images.githubusercontent.com/23142906/146753356-96f1b50f-11aa-47a0-b886-de4d94d50814.png)

acicovic committed:
Settings page: Show error message when duplicate tracking selected

If the user selects the same element in the "Post Types to Track" and "Page Types to Track" sections, an error message will be shown to them telling them that the element should be selected only in one of the sections.

The reasoning behind this is that we should not allow duplicate tracking.

acicovic committed:
Add comment to make code block purpose clearer

pauarge reviewed (changes requested):
Great work, but I'm leaving a few notes:

- Could we add (at least) one integration test and an end to end test to double check this functionality is working correctly?
- It would be great if you added a screenshot of the error in the PR's description.
- If it doesn't complicate the code very much, could we display which item(s) is duplicated in the error message? That would make the issue easier to understand for an end user.

acicovic commented:
> Could we add (at least) one integration test and an end to end test to double check this functionality is working correctly?

Will look into this.
> It would be great if you added a screenshot of the error in the PR's description.

Done! If anyone has an idea for a better error message, please let me know.

> If it doesn't complicate the code very much, could we display which item(s) is duplicated in the error message? That would make the issue easier to understand for an end user.

This isn't complicated to do and I thought of it myself, but was thinking that it could give some trouble to localized strings. But I should probably look into this more.

GaryJones commented:
> localized strings

You'll need to use variables (and add `/* translator: ... */` comments), but that's all standard and [documented](https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#variables).

That said, I'd rather we addressed the root issue as per https://github.com/Parsely/wp-parsely/pull/482#issuecomment-963287549 than simply rely on error notifications being noticed and acted upon.

acicovic commented:
Thanks for the input @GaryJones, I'll make sure to check the documentation you sent.

Concerning the root issue, it's on our radar and (from my understanding) monitored [here](https://github.com/Parsely/wp-parsely/issues/483). I'll bring your comment there for context, and we can resume conversation from there.

This PR doesn't replace that proposal. Rather, it's a relatively fast way to stop users from doing something they shouldn't do and the system shouldn't allow them to do. Regardless of the UI we'll finally implement, we should also be needing server-side validation in case anything else goes wrong.

pauarge commented:
> we should also be needing server-side validation in case anything else goes wrong.

This is especially important, I think this PR is definitely positive.

acicovic committed:
Mention duplicate post types in error message

acicovic committed:
Include integration test with duplicate tracking detection

acicovic commented:
I've just converted this PR to draft as it still needs some work and I've uncovered a bug.

acicovic committed:
Fix: don't save invalid tracking settings

Although an error message was being displayed for duplicate tracking, the specific settings were being saved into the database. This fix corrects that.

acicovic committed:
Fix and simplify settings page integration tests

acicovic commented:
Since we will be migrating to https://github.com/Parsely/wp-parsely/issues/483, we've agreed with @pauarge that an end-to-end test is not needed for this since it will change anyway.

So I'm flagging this for review.

GaryJones commented on src/UI/class-settings-page.php:
```diff
@@ -638,12 +638,19 @@ public function validate_options( array $input ): array {
-		$duplicate_tracking_found = ! empty( array_intersect( $input['track_post_types'], $input['track_page_types'] ) );
-		if ( $duplicate_tracking_found ) {
+		$duplicate_items = array_intersect( $input['track_post_types'], $input['track_page_types'] );
+		if ( ! empty( $duplicate_items ) ) {
```
Instead of the generic `empty()`, can we use a more specific `0 === count( $duplicate_items )`?

https://beberlei.de/2021/02/19/when_to_use_empty_in_php_i_say_never.html

acicovic replied:
I'm certainly not against it.

Looking at the current codebase of the plugin, we use `empty()` in a variety of places, which brings up the question of whether a refactor should be done at some point. In general I would say that `empty()` is more readable for many people as well as shorter, but we could go with the alternatives if we want to aim on better quality code.

For the time being, I'll post a commit addressing this.

GaryJones commented on src/UI/class-settings-page.php:
```diff
@@ -638,12 +638,19 @@ public function validate_options( array $input ): array {
+						'%s cannot be tracked as both posts and pages. Please select every element only once.',
+						'wp-parsely'
+					),
+					implode( ', ', $duplicate_items )
```
Can this be further refined to swap the last `, ` for an ` and `?

Right now it would say: "Foo, Bar cannot be tracked..." or "Foo, Bar, Baz" cannot be tracked...". Ideally, we could reframe it as "Foo and Bar cannot be tracked..." and "Foo, Bar and Baz cannot be tracked...".

Note also that some languages may handle this differently (or may not use `,` as comma punctuation at all), so correct i18n support may need a bit more thinking.

acicovic replied:
We would indeed need more tinkering for real/correct i18n support. Ideally, I would add post types in quotes which I think could make the message even more clear:

> "Foo", "Bar" and "Baz" cannot be tracked...

However, from the 3 languages I know, each one handles quotes differently. So for the time being I stepped back from this. For the time being I'll post a commit to swap the `','` with `and` and we can open a discussion about correct i18n support with the others.

GaryJones replied:
In terms of i18n support, I think swapping all commas for an "and" is going to be worse. I'll have a think about this.

acicovic committed:
Implement code review suggestions

acicovic commented:
@GaryJones I've marked the discussions as resolved but feel free to reopen anything if needed.

acicovic commented:
That `0 === count( $duplicate_items )` broke the tests so I'll look into it.

acicovic committed:
Fix bad comparison code

pauarge committed:
Merge branch 'develop' into add/settings-page-duplicate-tracking-validation

pauarge commented on tests/Integration/UI/SettingsPageTest.php:
```diff
@@ -0,0 +1,91 @@
+use Parsely\UI\Settings_Page;
+
+/**
+ * Row actions tests.
```
I think this might be a leftover from copy pasting, but this tests are not testing row actions.

acicovic replied:
This needs to be there (we use `self::$settings_page->validate_options()`).

pauarge replied:
I mean the `Row actions tests.` comment. It should be `Settings page tests.`

acicovic replied:
You are completely right. My eye didn't catch this.

pauarge commented on tests/Integration/UI/SettingsPageTest.php:
```diff
@@ -0,0 +1,91 @@
+		// Duplicate selection in Post Types.
+		$options['track_post_types'] = array( 'post', 'page' );
+		$options['track_page_types'] = array( 'page' );
+		self::assertSame( self::$settings_page->validate_options( $options ), $expected );
```
When asserting something, I think it's more readable to put first the `$expected` variable and then the computed one. Even better, we could separate the call to `validate_options` in a different line within a variable.

This comment applies for the other assertions as well.

GaryJones replied:
Agreed on `$expectation` coming first: https://phpunit.readthedocs.io/en/9.5/assertions.html#assertsame

acicovic replied:
This should be OK now. I've implemented both suggestions.

pauarge commented on tests/Integration/UI/SettingsPageTest.php:
```diff
@@ -0,0 +1,91 @@
+		// Duplicate selection in Page Types.
+		$options['track_post_types'] = array( 'post' );
+		$options['track_page_types'] = array( 'post', 'page' );
+		self::assertSame( self::$settings_page->validate_options( $options ), $expected );
```
I would separate the three assertions on three separate tests. In the event of a test failing, we would get better granularity.

pauarge reviewed (changes requested):
Great work! I've left some minor improvements in the test.

pauarge committed:
Merge branch 'develop' into add/settings-page-duplicate-tracking-validation

acicovic committed:
SettingsPageTest: Add expected values first in assertions

acicovic committed:
SettingsPageTest: Move validate_options() calls for readability

pauarge committed:
Merge branch 'develop' into add/settings-page-duplicate-tracking-validation

acicovic committed:
SettingsPageTest: Put every assertion in its own test

acicovic committed:
Merge branch 'add/settings-page-duplicate-tracking-validation' of https://github.com/Parsely/wp-parsely into add/settings-page-duplicate-tracking-validation

acicovic committed:
SettingsPageTest: Fix incorrect class comment

pauarge reviewed (approved):
LGTM, great work!

acicovic committed:
Merge branch 'develop' into add/settings-page-duplicate-tracking-validation
